lib/ecto/migration/index.ex

```diff
@@ -0,0 +1,192 @@
+defmodule Ecto.Migration.Index do
+  import Ecto.Query
+
+  defmacro __using__(_opts) do
+    quote do
+      import Ecto.Migration.Index
+      @indexes []
+      @before_compile Ecto.Migration.Index
+    end
+  end
+
+  defmacro index(columns, opts \\ []) do
+    opts = get_index_helper(opts, :fields_names) |> :lists.sort
+    quote do
+      @indexes [{unquote(List.wrap(columns)), unquote(opts)} | @indexes]
+    end
+  end
+
+  defmacro __before_compile__(_env) do
+    quote do
+      def get_indexes do
+        @indexes
+      end
+    end
+  end
+
+  def create_index(_, [], _) do
+    ""
+  end
+
+  def create_index(_, _, false) do
+    ""
+  end
+
+  def create_index(module, all_indexes, true) do
+    for {fields, opts} <- all_indexes do
+      tablename =  module.__schema__(:source) |> String.to_atom
+      quote do
+        create index(unquote(tablename), unquote(fields), unquote(opts))
+      end
+    end
+  end
+
+  def delete_index(false, _, _, _) do
+    ""
+  end
+
+  def delete_index(true, tablename, module, repo) do
+    fields_from_db = Ecto.Migration.Auto.get_existings_indexes(repo, tablename)
+    case fields_from_db do
+      [] ->
+        ""
+      _ ->
+        deletion = for db_index <- fields_from_db do
+          old_index_fields = index_string_to_atom_list(db_index)
+          old_index_opts = [index_name: db_index.index_name,
+                            concurrently: db_index.concurrently,
+                            index_type: db_index.index_type,
+                            unique: db_index.unique]
+          quote do
+            drop index(unquote(tablename |> String.to_atom), unquote(old_index_fields), unquote(old_index_opts))
+          end
+        end
+        # remove all old indexes from the system table
+        from(i in Ecto.Migration.SystemTable.Index, where: i.tablename == ^tablename) |> repo.delete_all
+        # insert new indexes to the system table
+        all_indexes = Ecto.Migration.Index.get_all_indexes(module)
+        for {fields, opts} <- all_indexes do
+          {index_name, concurently, unique, index_type} = Ecto.Migration.Index.get_index_helper(opts, [])
```

Comment:
Better alias Ecto.Migration.SystemTable.Index

Instead of index_helper, use simply:

```elixir
Enum.into(opts, %Index{tablename: tablename, index: Enum.join(fields, ",")})
```